TestHelpers/struct_field_to_index_map_generator.py

```diff
@@ -0,0 +1,312 @@
+    struct my_struct B;
+}
+'''
+# this C code will generate the following map:
+'''
+[A.x] => [A.0]
+[A.y] => [A.1]
+[B.x] => [B.0]
+[B.y] => [B.1]
+'''
+# this is used for Testers/Passes when we want to specify
+# a field of a variable by field name (rather than index)
+# to test its MustCall/CalledMethods
+
+import sys
+
+DEBUG = True
+
+
+def logout(x: str):
+    if (DEBUG):
+        print(x)
+
+
+class Field:
+    def __init__(self, field_name: str, field_index: int):
+        self.field_name = field_name
+        self.field_index = field_index
+
+
+class Struct:
+    def __init__(self, name: str):
+        self.name = name
+        self.typedefs: list[str] = []
+        self.fields: list[Field] = []
+
+    def get_name(self) -> str:
+        return self.name
+
+    def get_typedefs(self) -> "list[str]":
+        return self.typedefs
+
+    def add_field(self, field: Field) -> None:
+        self.fields.append(field)
+
+    def add_typedef(self, typedef: str) -> None:
+        self.typedefs.append(typedef)
+
+
+class StructsManager:
+    def __init__(self):
+        self.structs: list[Struct] = []
+
+    # if struct is not found, a new one is returned with name struct_name if error_if_not_found is false
+    def get_struct(self, struct_name: str, error_if_not_found: bool = False) -> Struct:
+        for struct in self.structs:
+            if struct.get_name() == struct_name or struct_name in struct.get_typedefs():
+                return struct
+
+        if error_if_not_found:
+            raise ValueError(f"struct {struct_name} not found")
+
+        new_struct = Struct(struct_name)
+        self.structs.append(new_struct)
+        return self.structs[-1]
+
+    def struct_exists(self, struct_name: str) -> bool:
+        for struct in self.structs:
+            if struct.get_name() == struct_name or struct_name in struct.get_typedefs():
+                return True
+
+        return False
+
+    # if there already exists a struct of name struct_name, that struct is returned.
+    # otherwise, the struct is added and returned.
+    def add_struct(self, struct_name: str) -> bool:
+        for struct in self.structs:
+            if struct.get_name() == struct_name:
+                return struct
+
+        new_struct = Struct(struct_name)
+        self.structs.append(new_struct)
+        return self.structs[-1]
+
+
+class StructVar:
+    def __init__(self, var_name: str, type_name: str):
+        self.var_name = var_name
+        self.type_name = type_name
+
+    def get_var_name(self):
+        return self.var_name
+
+    def get_type_name(self):
+        return self.type_name
+
+
+class StructVarManager:
+    def __init__(self):
+        self.struct_vars: list[StructVar] = []
+
+    # if there already exists a var of name var_name, that struct is returned.
+    # otherwise, the var is added and returned.
+    def add_struct_var(self, var_name: str, type_name: str) -> bool:
+        for struct_var in self.struct_vars:
```

Comment:
When I run `pylint` on this file, I get a large number of [W0621 warnings](https://pylint.readthedocs.io/en/latest/user_guide/messages/warning/redefined-outer-name.html) about redefining variable names. These should probably all be fixed, since they can be nasty to debug. Here's the full list:
```
struct_field_to_index_map_generator.py:43:40: W0621: Redefining name 'field_index' from outer scope (line 236) (redefined-outer-name)
struct_field_to_index_map_generator.py:60:24: W0621: Redefining name 'field' from outer scope (line 307) (redefined-outer-name)
struct_field_to_index_map_generator.py:122:12: W0621: Redefining name 'struct_var' from outer scope (line 301) (redefined-outer-name)
struct_field_to_index_map_generator.py:131:33: W0621: Redefining name 'structs_manager' from outer scope (line 239) (redefined-outer-name)
struct_field_to_index_map_generator.py:138:31: W0621: Redefining name 'field_index' from outer scope (line 236) (redefined-outer-name)
struct_field_to_index_map_generator.py:152:4: W0621: Redefining name 'field' from outer scope (line 307) (redefined-outer-name)
struct_field_to_index_map_generator.py:156:34: W0621: Redefining name 'struct_var_manager' from outer scope (line 240) (redefined-outer-name)
struct_field_to_index_map_generator.py:156:72: W0621: Redefining name 'structs_manager' from outer scope (line 239) (redefined-outer-name)
struct_field_to_index_map_generator.py:193:22: W0621: Redefining name 'structs_manager' from outer scope (line 239) (redefined-outer-name)
```

Reply:
(pylint also gives a bunch of warnings that aren't very helpful about style, but these ones are worth paying attention to)